Dockerfile

```diff
@@ -1,23 +1,30 @@
-FROM python:3.7-slim-buster AS builder
+FROM python:3.9-slim-buster AS builder
 
 RUN apt-get update -qq \
     && DEBIAN_FRONTEND=noninteractive apt-get install -qq -y  \
-        gcc \
+        build-essential \
         libgeos-dev \
+        apache2 \
+        apache2-dev \
     && apt-get clean \
     && rm -rf /var/lib/apt/lists/* \
-    && pip3 install --user pipenv
+    && pip3 install --user --no-warn-script-location --upgrade pip \
+    && pip3 install --user --no-warn-script-location pipenv
 
 ENV PIPENV_VENV_IN_PROJECT=1
 
 # Pipfile contains requests
-ADD Pipfile.lock Pipfile /usr/src/
+COPY Pipfile.lock Pipfile /usr/src/
 WORKDIR /usr/src
 
 # Create virtualenv with all dependencies
-RUN /root/.local/bin/pipenv sync
+# Install latest mod_wsgi using the pip command, this will automatically download, configure and
+# build mod_wsgi apache module inside the virtual environment. We do this separately and not part
+# of the Pipfile because we only need mod_wsgi on docker image and not locally.
+RUN /root/.local/bin/pipenv sync \
+    && /root/.local/bin/pipenv run pip install mod_wsgi
```

Comment:
@boecklic there are 3 possibles solutions:
1. Put `mod_wsgi` in a separate layer
    
    ```dockerfile
    RUN /root/.local/bin/pipenv run pip install mod_wsgi
    COPY Pipfile.lock Pipfile /usr/src/
    RUN /root/.local/bin/pipenv sync
    ```

3. Put `mod_wsgi` with the apt layer

    ```dockerfile
    RUN apt-get update -qq \
        && DEBIAN_FRONTEND=noninteractive apt-get install -qq -y  \
           build-essential \
           libgeos-dev \
           apache2 \
           apache2-dev \
       && apt-get clean \
       && rm -rf /var/lib/apt/lists/* \
       && pip3 install --user --no-warn-script-location --upgrade pip \
       && pip3 install --user --no-warn-script-location pipenv \
       && /root/.local/bin/pipenv run pip install mod_wsgi 

    # Pipfile contains requests
    COPY Pipfile.lock Pipfile /usr/src/

    # Create virtualenv with all dependencies
    RUN /root/.local/bin/pipenv sync
    ```

5. Put `mod_wsgi` with the other python virtualenv layer (current solution)

I would prefer solution 3. (current) over 2. because the `apt` layer install tools for building the virtual environment and `pipenv run pip install mod_wsgi` install `mod_wsgi` inside the virtual environment, so it seem more logical to keep related stuff in the same layer. In solution 3. the apt layer is completely discarded in the runtime build stage which only copy the virtual env of the builder stage. I think solution 3. is clearer over 2. about what is taken from builder into runtime stage and it keep related stuff together.

We could take solution 1. over 3. but our best practice is to keep the number of layer as low as possible and I don't see a big improvement separating it in a new layer.

What do you think ?

Reply:
Jup, keep the current solution!